### PR TITLE
BUGFIX: RAIL-1469 SDK: Can't add Rollup (totals) on visualization component

### PR DESCRIPTION
**Pull request by tomasreichmann**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [ ] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-app-components:
- gdc-analytical-designer:
- gdc-dashboards:

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3470/ : SUCCESS in 8m 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4298/ : SUCCESS in 6s

File: src/components/uri/Visualization.tsx [hunk 1]
```diff
@@ -368,16 +346,32 @@ export class VisualizationWrapped extends React.Component<
         switch (type) {
             case VisualizationTypes.TABLE:
                 return <TableComponent {...commonProps} totals={totals} />;
-            case VisualizationTypes.PIVOT_TABLE:
-                return <PivotTableComponent {...commonProps} totals={totals} />;
+            case VisualizationTypes.PIVOT_TABLE: {
+                const pivotBucketProps = mdObjectToPivotBucketProps(mdObject, filtersFromProps);
+                const pivotTableProps: IPivotTableProps = {
```
Comment (no23reason) [inline src/components/uri/Visualization.tsx]: This is just `commonProps` without `resultSpec`, right? Maybe you could simplify that?

File: src/components/uri/Visualization.tsx [hunk 2]
```diff
@@ -411,25 +407,25 @@ export class VisualizationWrapped extends React.Component<
...
               }
             : {};
-        const resultSpecWithDimensionsAndSorting = {
+        const resultSpec = {
             ...resultSpecWithDimensions,
             ...treemapDefaultSorting,
         };
 
-        const dataSource = await this.adapter.createDataSource(afmWithFilters);
+        const dataSource = await this.adapter.createDataSource(afm, JSON.stringify(afm));
```
Comment (no23reason) [inline src/components/uri/Visualization.tsx]: I believe `JSON.stringify(afm)` is not needed here, as the default does som kind of `stringify` as well.

**Comment by tomasreichmann** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-955.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

File: src/helpers/MdObjectHelper.ts [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
 import get = require("lodash/get");
-import { VisualizationObject } from "@gooddata/typings";
+import { VisualizationObject, AFM, VisualizationInput } from "@gooddata/typings";
 import { IMeasureTitleProps, IArithmeticMeasureTitleProps } from "..";
-import { ATTRIBUTE } from "../constants/bucketNames";
+import { ATTRIBUTE, MEASURES, COLUMNS } from "../constants/bucketNames";
 import IMeasure = VisualizationObject.IMeasure;
 import IArithmeticMeasureDefinition = VisualizationObject.IArithmeticMeasureDefinition;
+import { IPivotTableBucketProps } from "../components/PivotTable";
+import { toAfmResultSpec } from "@gooddata/gooddata-js/lib/DataLayer/converters/toAfmResultSpec";
```
Comment (no23reason) [inline src/helpers/MdObjectHelper.ts]: Nitpick: this import should be nex to other "package" imports

File: src/helpers/MdObjectHelper.ts [hunk 4]
```diff
@@ -54,9 +57,44 @@ function buildArithmeticMeasureTitleProps(
...
+    const measureBucket = mdObject.content.buckets.find(bucket => bucket.localIdentifier === MEASURES);
+    const rowBucket = mdObject.content.buckets.find(bucket => bucket.localIdentifier === ATTRIBUTE);
+    const columnBucket = mdObject.content.buckets.find(bucket => bucket.localIdentifier === COLUMNS);
+
+    const measures: IPivotTableBucketProps["measures"] = (measureBucket && measureBucket.items) || [];
+    const rows: IPivotTableBucketProps["rows"] =
+        (rowBucket && (rowBucket.items as VisualizationObject.IVisualizationAttribute[])) || [];
+    const columns: IPivotTableBucketProps["columns"] =
+        (columnBucket && (columnBucket.items as VisualizationObject.IVisualizationAttribute[])) || [];
+    const sortBy: IPivotTableBucketProps["sortBy"] = JSON.parse(mdObject.content.properties).sortItems || [];
```
Comment (no23reason) [inline src/helpers/MdObjectHelper.ts]: `JSON.parse` can throw if the input is invalid. Is that ok here?

File: src/helpers/tests/MdObjectHelper.spec.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // (C) 2007-2018 GoodData Corporation
 import MdObjectHelper from "../MdObjectHelper";
-import { visualizationObjects } from "../../../__mocks__/fixtures";
-import { VisualizationObject } from "@gooddata/typings";
+import { visualizationObjects, pivotTableMDO } from "../../../__mocks__/fixtures";
+import { VisualizationObject, AFM, VisualizationInput } from "@gooddata/typings";
```
Comment (no23reason) [inline src/helpers/tests/MdObjectHelper.spec.ts]: Nitpick: import order

File: src/components/uri/Visualization.tsx [hunk 6]
```diff
@@ -395,13 +389,15 @@ export class VisualizationWrapped extends React.Component<
 
         const visualizationClass = await fetchVisualizationClass(this.sdk, visualizationClassUri);
         const processedVisualizationObject = fillMissingTitles(mdObject.content, locale);
-        const { afm, resultSpec } = toAfmResultSpec(processedVisualizationObject);
 
+        // getTotals might ignore subTotals
```
Comment (no23reason) [inline src/components/uri/Visualization.tsx]: What are we going to do about that?

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1192/ : SUCCESS in 36m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4300/ : SUCCESS in 11s

File: src/components/uri/Visualization.tsx (hunk 1 above)
Comment (tomasreichmann) [reply src/components/uri/Visualization.tsx]: updated

File: src/components/uri/Visualization.tsx (hunk 6 above)
Comment (tomasreichmann) [reply src/components/uri/Visualization.tsx]: It was just my comment, but there is a separate ticket for subtotals, to I leave this to BB

Comment (tomasreichmann) [reply src/components/uri/Visualization.tsx]: I removed the comment

File: src/components/uri/Visualization.tsx (hunk 2 above)
Comment (tomasreichmann) [reply src/components/uri/Visualization.tsx]: ok, removed

File: src/helpers/MdObjectHelper.ts (hunk 3 above)
Comment (tomasreichmann) [reply src/helpers/MdObjectHelper.ts]: Changed to import { DataLayer } from "@gooddata/gooddata-js";
DataLayer. toAfmResultSpec and moved

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3472/ : SUCCESS in 12m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4303/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build in pipeline gate aborted.

**Comment by tomasreichmann** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-955.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3477/ : SUCCESS in 3m 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4308/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1196/ : SUCCESS in 30m 05s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4309/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1861/ : SUCCESS in 16m 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3491/ : SUCCESS in 11m 49s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/955/ : SUCCESS in 34s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1026/ : SUCCESS in 31m 05s

**Pull request merged**